lib/Capabilities.php

```diff
@@ -0,0 +1,103 @@
+<?php
+/**
+ * @author Jan Ackermann <jackermann@owncloud.com>
+ *
+ * @copyright Copyright (c) 2018, ownCloud GmbH
+ * @license AGPL-3.0
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\PasswordPolicy;
+
+use OCP\Capabilities\ICapability;
+use OCP\IConfig;
+
+/**
+ * Class Capabilities
+ *
+ * @package OCA\PasswordPolicy
+ */
+class Capabilities implements ICapability {
+
+	/** @var IConfig */
+	private $config;
+
+	/**
+	 * Capabilities constructor.
+	 *
+	 * @param IConfig $config
+	 */
+	public function __construct(IConfig $config) {
+		$this->config = $config;
+	}
+
+	/**
+	 * Return this classes capabilities
+	 *
+	 * @return array
+	 */
+	public function getCapabilities() {
+		$generalPasswordRequirements = ['configuration' => [
+			'lower_case' => [
+				"generate_from" => ["characters" => "abcdefghijklmnopqrstuvwxyz"],
+			],
+			'upper_case' => [
+				"generate_from" => ["characters" => "ABCDEFGHIJKLMNOPQRSTUVWXYZ"],
+			],
+			'numbers' => [
+				"generate_from" => ["characters" => "1234567890"],
+			],
+			'special_characters' => [
+				"generate_from" => []
+			]],
+		];
+
+		if ($this->config->getAppValue('password_policy', 'spv_min_chars_checked') === 'on') {
```

Comment:
This seems dense...
Maybe you can split the data so you could loop over it, since the code is very similar
```
$data = [
  [
    'config_key_check' => 'spv_min_chars_checked',
    'config_key_value' => 'spv_min_chars_value',
    'destination' => ['minimum_characters'],
  ],
  [
    'config_key_check' => 'spv_lowercase_checked',
    'config_key_value' => 'spv_lowercase_value',
    'destination' => ['configuration','lower_case','minimum'],
  ],
];
for ($data as $item) {
  if ($this->config->getAppValue('password_policy', $item['config_key_check']) === 'on') {
    $value = \intval($this->config->getAppValue('password_policy', $item['config_key_value']));
    $target =& $generalPasswordRequirements;
    foreach ($item['destination'] as $key) {
      $target =& $target[$key];
    }
    $target = $value;
    unset($target);
  }
}
```
The problem there is that playing with references is risky. You could also try to move the assignment part to a function and pass the target array by reference
```
for ($data as $item) {
  if ($this->config->getAppValue('password_policy', $item['config_key_check']) === 'on') {
    $value = \intval($this->config->getAppValue('password_policy', $item['config_key_value']));
    $this->assignTo($generalPasswordRequirements, $item['destination'], $value);
  }
}
```

Reply:
I don't like this 'design' approach, from my POV, this makes it unnecessary complicated to read.

Reply:
Can we reuse the configProvider here? something like:
```
if (($spv_min_chars = $this->configProvider->get('spv_min_chars')) !== null) {
  $generalPasswordRequirements['minimum_characters'] = $spv_min_chars;
}
```

Reply:
For sure!

Reply:
@jvillafanez should the configProvider also coerce the values to the matching integer Datatype?

Reply:
Probably not because it can't guarantee the same type for everything. It's possible at the moment, but there is no guarantee it will remain the same.